### DISCUSSION TITLE
'nounique' doesn't seem to work

### DISCUSSION THREAD

**Discussion by So1ene** [gulpjs/gulp#2677] [Help]
I want to use gulp to concat some html files in order, and some of them are duplicates.
I'm using 'nounique' to be able to have multiple of the same src

For example:

```

gulp.task('Generate HTML', function(cb) {
  const srcset = [
    'header.html',
    'content.html',
    'columns-left.html',
    'columns-right.html',
    'columns-left.html',
    'columns-right.html',
    'banner.html',
    'content.html',
    'content.html',
    'banner.html',
    'footer.html'
  ];

  gulp.src(srcset, {'allowEmpty': true, 'nounique': true, 'nosort': true})
    .pipe(concat('generated-page.html'))
    .pipe(gulp.dest('public/'));
    
  cb();
});

```

I found the options `no-unique` and `no-sort` in the documentation so I thought that would solve my problem, however I am not getting the desired result, as every html snippet is only `concat`-ed one time, so any duplicated snippets will only show up once.

I'm not sure if this is because of src or because of concat, please let me know if you can help me

**Comment by So1ene** [comment]
So I seem to have found the answer in case anyone is interested thanks to some help on stackOverflow:

"
gulp.src treats its argument as a list of glob patterns rather than file paths. Each pattern could match one or multiple files or none at all, and if the same file is matched by multiple patterns, it's up to the implementation to decide what to do.

It's better here to load each file individually and then join the streams.

To use [merge-stream](https://www.npmjs.com/package/merge-stream), install the package with:

npm install -D merge-stream
Then your code could look like this:

```
const mergeStream = require('merge-stream');

gulp.task('Generate HTML', function(cb) {
  const srcset = [
    'header.html',
    'content.html',
    'columns-left.html',
    'columns-right.html',
    'columns-left.html',
    'columns-right.html',
    'banner.html',
    'content.html',
    'content.html',
    'banner.html',
    'footer.html'
  ];

  const streams = srcset.map(src => gulp.src(src));
  mergeStream(...streams)
    .pipe(concat('generated-page.html'))
    .pipe(gulp.dest('public/'));
    
  cb();
});
```

**Reply by phated** [reply]
Calling `cb` while using streams is wrong.

**Reply by So1ene** [reply]
I’m new to this, can you explain why?

**Reply by phated** [reply]
You should read https://gulpjs.com/docs/en/getting-started/async-completion - only 1 type of async should be used in any task

**Reply by So1ene** [reply]
Thanks, Im now using promises instead

**Reply by phated** [reply]
Promises + streams is **still** wrong. You need to use exactly 1 type of async.

**Reply by So1ene** [reply]
Ohhh ok so i just need to return the stream
